eth2near/utilities/src/eth_proof_generator.rs

```diff
@@ -0,0 +1,178 @@
+    let hasher = Arc::new(HasherKeccak::new());
+    let mut trie = PatriciaTrie::new(Arc::clone(&memdb), Arc::clone(&hasher));
+
+    for (_, receipt) in receipts.iter().enumerate() {
+        let receipt_key = rlp::encode(&receipt.transaction_index).to_vec();
+        let receipt_data = encode_receipt(receipt);
+
+        trie.insert(receipt_key, receipt_data)?;
+    }
+
+    Ok(trie)
+}
+
+fn encode_receipt(receipt: &TransactionReceipt) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+
+    if receipt.transaction_type != 0 {
+        stream.append(&receipt.transaction_type);
+    }
+
+    stream.begin_list(4);
+    stream
+        .append(&receipt.status)
+        .append(&receipt.cumulative_gas_used.0)
+        .append(&receipt.logs_bloom.0.to_vec());
+
+    stream.begin_list(receipt.logs.len());
+    for (_, log) in receipt.logs.iter().enumerate() {
+        stream.begin_list(3);
+        stream.append(&log.address.0.to_vec());
+
+        stream.begin_list(log.topics.len());
+        for (_, topic) in log.topics.iter().enumerate() {
+            stream.append(&topic.0.to_vec());
+        }
+
+        stream.append(&log.data.0);
+    }
+
+    stream.out().to_vec()
+}
+
+fn encode_log(log: &Log) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+    stream.begin_list(3);
+
+    stream.append(&log.address.0.to_vec());
+
+    stream.begin_list(log.topics.len());
+    for (_, topic) in log.topics.iter().enumerate() {
+        stream.append(&topic.0.to_vec());
+    }
+
+    stream.append(&log.data.0);
+
+    stream.out().to_vec()
+}
+
+fn encode_header(header: &BlockHeader) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+    stream.begin_unbounded_list();
+
+    stream
+        .append(&header.parent_hash.0.to_vec())
+        .append(&header.sha3_uncles.0.to_vec())
+        .append(&header.miner.0.to_vec())
+        .append(&header.state_root.0.to_vec())
+        .append(&header.transactions_root.0.to_vec())
+        .append(&header.receipts_root.0.to_vec())
+        .append(&header.logs_bloom.0.to_vec())
+        .append(&header.difficulty)
+        .append(&header.number.0)
+        .append(&header.gas_limit.0)
+        .append(&header.gas_used.0)
+        .append(&header.timestamp.0)
+        .append(&header.extra_data.0.to_vec())
+        .append(&header.mix_hash.0.to_vec())
+        .append(&header.nonce.0.to_vec());
+
+    if header.base_fee_per_gas.is_some() {
+        stream.append(&header.base_fee_per_gas.unwrap());
+    }
+
+    if header.withdrawals_root.is_some() {
+        stream.append(&header.withdrawals_root.clone().unwrap().0.to_vec());
+    }
+
+    if header.blob_gas_used.is_some() {
+        stream.append(&header.blob_gas_used.unwrap());
+    }
+
+    if header.excess_blob_gas.is_some() {
+        stream.append(&header.excess_blob_gas.unwrap());
+    }
+
+    if header.parent_beacon_block_root.is_some() {
+        stream.append(&header.parent_beacon_block_root.clone().unwrap().0.to_vec());
+    }
```

Comment:
```suggestion
    if let Some(base_fee_per_gas) = header.base_fee_per_gas {
        stream.append(&base_fee_per_gas);
    }

    if let Some(withdrawals_root) = header.withdrawals_root.as_ref() {
        stream.append(&withdrawals_root.0.as_slice());
    }

    if let Some(blob_gas_used) = header.blob_gas_used {
        stream.append(&blob_gas_used);
    }

    if let Some(excess_blob_gas) = header.excess_blob_gas {
        stream.append(&excess_blob_gas);
    }

    if let Some(parent_beacon_block_root) = header.parent_beacon_block_root.as_ref() {
        stream.append(&parent_beacon_block_root.0.as_slice());
    }
```

Reply:
We can write some macros to append optional fields to reduce the number of `if` statements. But it is not so important, so let's do that in separate PR.

Reply:
Fixed in 6a2e37708ec5a72cc0dc99b7634f0cc796ea9278